query-service-impl/src/main/java/org/hypertrace/core/query/service/postgres/PostgresResultAnalyzer.java

```diff
@@ -0,0 +1,75 @@
+package org.hypertrace.core.query.service.postgres;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.RateLimiter;
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Discovers the map attributes indexes from Postgres Result Set */
+class PostgresResultAnalyzer {
+  private static final Logger LOG = LoggerFactory.getLogger(PostgresResultAnalyzer.class);
+
+  /* Stores the Non-Map Attributes logical name to Physical Name index */
+  private final Map<String, Integer> logicalNameToPhysicalNameIndex;
+  private final ResultSet resultSet;
+  private final Map<String, RateLimiter> attributeLogRateLimitter;
+
+  PostgresResultAnalyzer(
+      ResultSet resultSet,
+      LinkedHashSet<String> selectedAttributes,
+      Map<String, Integer> logicalNameToPhysicalNameIndex) {
+    this.logicalNameToPhysicalNameIndex = logicalNameToPhysicalNameIndex;
+    this.resultSet = resultSet;
+    this.attributeLogRateLimitter = new HashMap<>();
+    selectedAttributes.forEach(e -> attributeLogRateLimitter.put(e, RateLimiter.create(0.5)));
+  }
+
+  /** For each selected attributes build the map of logical name to result index. */
+  static PostgresResultAnalyzer create(
+      ResultSet resultSet,
+      LinkedHashSet<String> selectedAttributes,
+      TableDefinition tableDefinition)
+      throws SQLException {
+    Map<String, Integer> logicalNameToPhysicalNameIndex = new HashMap<>();
+
+    for (String logicalName : selectedAttributes) {
+      String name = tableDefinition.getPhysicalColumnName(logicalName);
+      ResultSetMetaData metaData = resultSet.getMetaData();
+      for (int colIndex = 0; colIndex < metaData.getColumnCount(); colIndex++) {
```

Comment:
we can create a map of physical column name to column index as preprocessed step, and use that map as lookup. 

e.g below will go outside of `selectedAttributes` loop
```
Map<String, Integer> physicalToIndex = new HashMap<>();
for (int colIndex = 0; colIndex < metaData.getColumnCount(); colIndex++) {
String physName = metaData.getColumnName(colIndex);
physicalToIndex.put(physName, colIndex)
}
```

Reply:
Makes sense. Done.